packages/@react-spectrum/s2/src/NotificationBadge.tsx

```diff
@@ -0,0 +1,146 @@
+import {AriaLabelingProps, DOMProps, DOMRef, DOMRefValue} from '@react-types/shared';
+import {ContextValue, SlotProps} from 'react-aria-components';
+import {filterDOMProps} from '@react-aria/utils';
+import {getAllowedOverrides, StyleProps} from './style-utils' with {type: 'macro'};
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import {NumberFormatter} from '@internationalized/number';
+import React, {createContext, forwardRef} from 'react';
+import {style} from '../style' with {type: 'macro'};
+import {useDOMRef} from '@react-spectrum/utils';
+import {useLocale, useLocalizedStringFormatter} from '@react-aria/i18n';
+import {useSpectrumContextProps} from './useSpectrumContextProps';
+
+export interface NotificationBadgeStyleProps {
+  /**
+   * The size of the notification badge.
+   *
+   * @default 'S'
+   */
+  size?: 'S' | 'M' | 'L' | 'XL'
+}
+
+export interface NotificationBadgeProps extends DOMProps, AriaLabelingProps, StyleProps, NotificationBadgeStyleProps, SlotProps {
+  /**
+   * The value to be displayed in the notification badge.
+   */
+  value?: number
+}
+
+export const NotificationBadgeContext = createContext<ContextValue<Partial<NotificationBadgeProps>, DOMRefValue<HTMLDivElement>>>(null);
+
+const badge = style({
+  display: 'flex',
+  font: 'control',
+  fontSize: {
+    size: {
+      S: 'ui-xs',
+      M: 'ui-xs',
+      L: 'ui-sm',
+      XL: 'ui'
+    }
+  },
+  justifyContent: 'center',
+  alignItems: 'center',
+  backgroundColor: 'accent',
+  height: {
+    size: {
+      S: {
+        default: 12,
+        isEmpty: 8
+      },
+      M: {
+        default: '[14px]',
+        isEmpty: 8
+      },
+      L: {
+        default: 16,
+        isEmpty: '[10px]'
+      },
+      XL: {
+        default: 18,
+        isEmpty: '[10px]'
+      }
+    }
+  },
+  aspectRatio: {
+    isEmpty: 'square',
+    isSingleDigit: 'square'
+  },
+  width: 'fit',
+  paddingX: {
+    isDoubleDigit: 'edge-to-text'
+  },
+  borderRadius: 'pill'
+}, getAllowedOverrides());
+
+/**
+ * Notification badges are used to indicate new or pending activity .
+ */
+export const NotificationBadge = forwardRef(function Badge(props: NotificationBadgeProps, ref: DOMRef<HTMLDivElement>) {
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-spectrum/s2');
+  [props, ref] = useSpectrumContextProps(props, ref, NotificationBadgeContext);
+  let {
+    size = 'S',
+    value,
+    ...otherProps
+  } = props;
+  let domRef = useDOMRef(ref);
+  let {locale} = useLocale();
+
+  if (value && value < 0) {
+    throw new Error('Value cannot be negative');
+  }
+
+  let truncatedValue: number | undefined = undefined;
+  if (value && value > 99) {
+    truncatedValue = 99;
+  }
+
+  let formattedValue = '';
+  if (truncatedValue) {
+    formattedValue = new NumberFormatter(locale).format(truncatedValue);
+  } else if (value) {
+    formattedValue = new NumberFormatter(locale).format(value);
+  }
+
+  let length = formattedValue.length;
```

Comment:
Might be better to use math to figure out number of digits just in case a number is multibyte character at some point
https://stackoverflow.com/questions/14879691/get-number-of-digits-with-javascript

Reply:
maybe im just overthinking this but im not sure if this would work for characters like 四 (4). it's technically a number but it's also more so a character so a formula like `Math.log(number) * Math.LOG10E + 1 | 0` (as suggested in the post) wouldn't work.

i was thinking we use the length of the value after it's been formatted, otherwise, we make the assumption that something like the number 20 in another locale will always convert to a number that is also 2 in length

Reply:
maybe im overthinking this but im not sure if this would work with a character like 四 (4) with the formula suggested in the post `var length = Math.log(number) * Math.LOG10E + 1 | 0`

i was thinking we should use the length of the formatted value, otherwise, are we assuming that a number like 20 in another locale will always equal a formatted string with a length of 2?

Reply:
We only support base 10 systems, so I believe that's a true statement. 20 requires one tens digit and one ones digit.

But maybe best to ask in the slack channel for g11n about the best way to determine this.